diff --git a/Makefile.am b/Makefile.am
index 9a6f82c3..d32b102d 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -590,6 +590,11 @@ builtin_sources += plugins/bluez5.c plugins/bluez5.h
 builtin_modules += hfp_ag_bluez5
 builtin_sources += plugins/hfp_ag_bluez5.c plugins/bluez5.h
 
+if SAILFISH_BT
+builtin_modules += sfos_bt
+builtin_sources += plugins/sailfish_bt.c
+endif
+
 if UPOWER
 builtin_modules += upower
 builtin_sources += plugins/upower.c
diff --git a/configure.ac b/configure.ac
index 510d2595..ba2856cc 100644
--- a/configure.ac
+++ b/configure.ac
@@ -238,6 +238,11 @@ AC_ARG_ENABLE(bluetooth, AS_HELP_STRING([--disable-bluetooth],
 
 AM_CONDITIONAL(BLUETOOTH, test "${enable_bluetooth}" != "no")
 
+AC_ARG_ENABLE(sailfish-bt, AS_HELP_STRING([--enable-sailfish-bt],
+			[enable Sailfish OS Bluetooth plugin]),
+					[enable_sailfish_bt=${enableval}])
+AM_CONDITIONAL(SAILFISH_BT, test "${enable_sailfish_bt}" = "yes")
+
 AC_ARG_ENABLE(upower, AS_HELP_STRING([--disable-upower],
 			[disable UPower plugin]),
 					[enable_upower=${enableval}])
