diff --git a/src/ofono.h b/src/ofono.h
index 5500b734..81e831b1 100644
--- a/src/ofono.h
+++ b/src/ofono.h
@@ -106,6 +106,12 @@ void __ofono_modules_cleanup(void);
 int __ofono_plugin_init(const char *pattern, const char *exclude);
 void __ofono_plugin_cleanup(void);
 
+void __ofono_plugin_foreach(void (*fn) (struct ofono_plugin_desc *desc,
+			int flags, void *user_data), void *user_data);
+
+#define OFONO_PLUGIN_FLAG_BUILTIN (0x01)
+#define OFONO_PLUGIN_FLAG_ACTIVE  (0x02)
+
 #include <ofono/modem.h>
 
 typedef void (*ofono_modem_foreach_func)(struct ofono_modem *modem,
diff --git a/src/plugin.c b/src/plugin.c
index ad4b4cdc..25961b08 100644
--- a/src/plugin.c
+++ b/src/plugin.c
@@ -85,6 +85,25 @@ static gboolean check_plugin(struct ofono_plugin_desc *desc,
 	return TRUE;
 }
 
+void __ofono_plugin_foreach(void (*fn) (struct ofono_plugin_desc *desc,
+				int flags, void *user_data), void *user_data)
+{
+	GSList *list;
+
+	for (list = plugins; list; list = list->next) {
+		struct ofono_plugin *plugin = list->data;
+		int flags = 0;
+
+		if (!plugin->handle)
+			flags |= OFONO_PLUGIN_FLAG_BUILTIN;
+
+		if (plugin->active)
+			flags |= OFONO_PLUGIN_FLAG_ACTIVE;
+
+		fn(plugin->desc, flags, user_data);
+	}
+}
+
 #include "builtin.h"
 
 int __ofono_plugin_init(const char *pattern, const char *exclude)
