diff --git a/src/manager.c b/src/manager.c
index f34fd330..03bbd829 100644
--- a/src/manager.c
+++ b/src/manager.c
@@ -86,6 +86,8 @@ static int manager_init(void)
 	DBusConnection *conn = ofono_dbus_get_connection();
 	gboolean ret;
 
+	DBG("manager_init");
+
 	ret = g_dbus_register_interface(conn, OFONO_MANAGER_PATH,
 					OFONO_MANAGER_INTERFACE,
 					manager_methods, manager_signals,
diff --git a/src/ofono.h b/src/ofono.h
index 5c7bf691..5ab3be9e 100644
--- a/src/ofono.h
+++ b/src/ofono.h
@@ -616,6 +616,14 @@ void __ofono_private_network_release(int id);
 ofono_bool_t __ofono_private_network_request(ofono_private_network_cb_t cb,
 						int *id, void *data);
 
+
+#include <ofono/dbus-access.h>
+#include <ofono/slot.h>
+
+void __ofono_slot_manager_init(void);
+void __ofono_slot_manager_cleanup(void);
+
+
 #include <ofono/cell-info.h>
 #include <ofono/dbus-access.h>
 #include <ofono/netmon.h>
diff --git a/src/slot-manager.c b/src/slot-manager.c
index 00c693fb..4d7857d5 100644
--- a/src/slot-manager.c
+++ b/src/slot-manager.c
@@ -1466,7 +1466,7 @@ static void ofono_slot_manager_object_class_init(OfonoSlotManagerObjectClass *c)
 
 static OfonoSlotManagerObject *slot_manager = NULL;
 
-void __ofono_slot_manager_init(void)
+int slot_manager_init(void)
 {
 	/*
 	 * Let's not assume what's called first, ofono_slot_driver_register()
@@ -1475,9 +1475,10 @@ void __ofono_slot_manager_init(void)
 	if (!slot_manager) {
 		slot_manager = ofono_slot_manager_new();
 	}
+	return 0;
 }
 
-void __ofono_slot_manager_cleanup(void)
+void slot_manager_cleanup(void)
 {
 	if (slot_manager) {
 		OfonoSlotManagerObject *mgr = slot_manager;
@@ -1496,6 +1497,8 @@ void __ofono_slot_manager_cleanup(void)
 	}
 }
 
+OFONO_MODULE(slot_manager, slot_manager_init, slot_manager_cleanup)
+
 /*==========================================================================*
  * Driver registration
  * Requires access to slot_manager variable
