diff --git a/src/ofono.service.in b/src/ofono.service.in
index c24ac280..c33c615e 100644
--- a/src/ofono.service.in
+++ b/src/ofono.service.in
@@ -1,12 +1,17 @@
 [Unit]
 Description=Telephony service
-After=syslog.target
+Requires=dbus.service
+After=dbus.service
 
 [Service]
 Type=dbus
 BusName=org.ofono
-ExecStart=@prefix@/sbin/ofonod -n
+User=root
+EnvironmentFile=-/var/lib/environment/ofono/*.conf
+ExecStart=@prefix@/sbin/ofonod -n $OFONO_ARGS $OFONO_DEBUG
 StandardError=null
+Restart=always
+RestartSec=3
 
 [Install]
 WantedBy=multi-user.target
