diff --git a/drivers/qmimodem/qmi.c b/drivers/qmimodem/qmi.c
index b8f46a5b..0b46947d 100644
--- a/drivers/qmimodem/qmi.c
+++ b/drivers/qmimodem/qmi.c
@@ -26,6 +26,7 @@
 #include <ell/ell.h>
 
 #include <ofono/log.h>
+#include <gatchat.h>
 
 #include "qmi.h"
 #include "ctl.h"
@@ -92,6 +93,7 @@ struct qmi_transport {
 	const struct qmi_transport_ops *ops;
 	struct debug_data debug;
 	bool writer_active : 1;
+	GAtChat *atmodem;
 };
 
 struct qmi_qmux_device {
@@ -930,6 +932,22 @@ bool qmi_qmux_device_has_service(struct qmi_qmux_device *qmux, uint16_t type)
 	return __qmux_service_info_find(qmux, type);
 }
 
+void qmi_device_set_atmodem(struct qmi_qmux_device *device, GAtChat* atmodem)
+{
+	if (device == NULL || device->transport == NULL)
+		return;
+
+	device->transport->atmodem = atmodem;
+}
+
+GAtChat* qmi_device_get_atmodem(struct qmi_service *service)
+{
+	if (service == NULL || service->family->transport == NULL)
+		return NULL;
+
+	return service->family->transport->atmodem;
+}
+
 static int qmi_qmux_device_write(struct qmi_transport *transport,
 					struct qmi_request *req)
 {
diff --git a/drivers/qmimodem/qmi.h b/drivers/qmimodem/qmi.h
index 0e48313e..3fd98100 100644
--- a/drivers/qmimodem/qmi.h
+++ b/drivers/qmimodem/qmi.h
@@ -7,6 +7,7 @@
 
 #include <stdbool.h>
 #include <stdint.h>
+#include <gatchat.h>
 
 #define QMI_SERVICE_CONTROL	0	/* Control service */
 #define QMI_SERVICE_WDS		1	/* Wireless data service */
@@ -59,6 +60,9 @@ typedef void (*qmi_destroy_func_t)(void *user_data);
 struct qmi_service;
 struct qmi_result;
 
+void qmi_device_set_atmodem(struct qmi_qmux_device *device, GAtChat* atmodem);
+GAtChat* qmi_device_get_atmodem(struct qmi_service *service);
+
 typedef void (*qmi_debug_func_t)(const char *str, void *user_data);
 typedef void (*qmi_qmux_device_discover_func_t)(void *user_data);
 typedef void (*qmi_qmux_device_create_client_func_t)(struct qmi_service *,
diff --git a/drivers/qmimodem/sms.c b/drivers/qmimodem/sms.c
index 3e2bef6e..f7482314 100644
--- a/drivers/qmimodem/sms.c
+++ b/drivers/qmimodem/sms.c
@@ -16,6 +16,8 @@
 #include <ofono/modem.h>
 #include <ofono/sms.h>
 
+#include <gatchat.h>
+
 #include "qmi.h"
 #include "wms.h"
 #include "util.h"
@@ -714,6 +716,15 @@ static void set_event_cb(struct qmi_result *result, void *user_data)
 	ofono_sms_register(sms);
 }
 
+static void qmi_sms_at_notify(GAtResult *result, gpointer user_data)
+{
+	struct ofono_sms *sms = user_data;
+
+	DBG("");
+
+	get_msg_list(sms);
+}
+
 static int qmi_sms_probe(struct ofono_sms *sms,
 				unsigned int vendor, void *user_data)
 {
@@ -740,6 +751,12 @@ static int qmi_sms_probe(struct ofono_sms *sms,
 
 	ofono_sms_set_data(sms, data);
 
+	GAtChat* atmodem = qmi_device_get_atmodem(wms);
+	if (atmodem) {
+		guint ret = g_at_chat_register(atmodem, "+CMTI:", qmi_sms_at_notify, FALSE, sms, NULL);
+		DBG("SMS AT CHAT REGISTER %d", ret);
+	}
+
 	return 0;
 }
 
diff --git a/drivers/qmimodem/voicecall.c b/drivers/qmimodem/voicecall.c
index aed662da..fe27502c 100644
--- a/drivers/qmimodem/voicecall.c
+++ b/drivers/qmimodem/voicecall.c
@@ -10,12 +10,15 @@
 #include <config.h>
 #endif
 
+#include <ofono.h>
 #include <ofono/log.h>
 #include <ofono/modem.h>
 #include <ofono/voicecall.h>
 #include <src/common.h>
 #include <ell/ell.h>
 
+#include <gatchat.h>
+
 #include "voice.h"
 
 #include "qmi.h"
@@ -707,6 +710,17 @@ static void send_dtmf(struct ofono_voicecall *vc, const char *dtmf,
 	send_one_dtmf(vc, *dtmf, send_one_dtmf_cb, vd);
 }
 
+static void qmi_voicecall_ring_notify(GAtResult *result, gpointer user_data) {
+	struct ofono_voicecall *vc = user_data;
+	struct voicecall_data *data = ofono_voicecall_get_data(vc);
+
+	DBG("DETECTED AT RING");
+
+	qmi_service_send(data->voice, QMI_VOICE_GET_ALL_CALL_INFO, NULL,
+				all_call_status_ind, vc, NULL);
+
+}
+
 static int qmi_voicecall_probe(struct ofono_voicecall *vc,
 					unsigned int vendor, void *user_data)
 {
@@ -724,6 +738,15 @@ static int qmi_voicecall_probe(struct ofono_voicecall *vc,
 
 	ofono_voicecall_set_data(vc, data);
 
+	//Get a handle to the modem serial interface to detect RING signals
+	//Sometimes the QMI notify doesnt come through if the phone is
+	//asleep
+	GAtChat* atmodem = qmi_device_get_atmodem(voice);
+	if (atmodem) {
+		guint ret = g_at_chat_register(atmodem, "RING", qmi_voicecall_ring_notify, FALSE, vc, NULL);
+		DBG("VOICE AT CHAT REGISTER %d", ret);
+	}
+
 	return 0;
 }
 
diff --git a/plugins/gobi.c b/plugins/gobi.c
index c1b30918..c77094a5 100644
--- a/plugins/gobi.c
+++ b/plugins/gobi.c
@@ -15,6 +15,7 @@
 #include <stdlib.h>
 #include <stdio.h>
 #include <net/if.h>
+#include <gatchat.h>
 
 #define OFONO_API_SUBJECT_TO_CHANGE
 #include <ofono/plugin.h>
@@ -47,6 +48,12 @@
 #include <drivers/qmimodem/wda.h>
 #include <drivers/qmimodem/util.h>
 
+//Define function from atutil.h as it cant be uncluded
+GAtChat *at_util_open_device(struct ofono_modem *modem, const char *key,
+				GAtDebugFunc debug_func, char *debug_prefix,
+				char *tty_option, ...);
+
+
 #define GOBI_DMS	(1 << 0)
 #define GOBI_NAS	(1 << 1)
 #define GOBI_WMS	(1 << 2)
@@ -614,6 +621,15 @@ static int gobi_enable(struct ofono_modem *modem)
 		qmi_qmux_device_set_io_debug(data->device,
 						gobi_io_debug, "QMI: ");
 
+	//Get a handle to an AT interface if one was found, and attach it to the qmi_device
+	GAtChat *atchat = at_util_open_device(modem, "Aux", NULL, "Aux: ", NULL);
+	if (!atchat) {
+		DBG("No Aux");
+	} else {
+		DBG("Have atmodem");
+	}
+	qmi_device_set_atmodem(data->device, atchat); //Will set to NULL if doesnt exist
+
 	r = qmi_qmux_device_discover(data->device, discover_cb, modem, NULL);
 	if (!r)
 		return -EINPROGRESS;
diff --git a/plugins/udevng.c b/plugins/udevng.c
index 1c56e2ea..b089a227 100644
--- a/plugins/udevng.c
+++ b/plugins/udevng.c
@@ -2149,6 +2149,7 @@ static struct {
 	{ "quectelqmi",	"qcserial",	"2c7c", "0121"	},
 	{ "quectelqmi",	"qmi_wwan",	"2c7c", "0125"	},
 	{ "quectelqmi",	"qcserial",	"2c7c", "0125"	},
+	{ "quectelqmi", "option",	"2c7c", "0125"	},
 	{ "quectelqmi",	"qmi_wwan",	"2c7c", "0195"	},
 	{ "quectelqmi",	"qcserial",	"2c7c", "0195"	},
 	{ "quectelqmi",	"option",	"2c7c", "0195"	},
@@ -2187,6 +2188,12 @@ static void check_usb_device(struct udev_device *device)
 	const char *syspath, *devname, *driver;
 	const char *vendor = NULL, *model = NULL;
 	const char *kernel_driver;
+	const char* devsyspath = udev_device_get_syspath(device);
+
+	if (devsyspath == NULL) {
+		return;
+	}
+	DBG("devsyspath: %s", devsyspath);
 
 	usb_device = udev_device_get_parent_with_subsystem_devtype(device,
 							"usb", "usb_device");
@@ -2258,10 +2265,39 @@ static void check_usb_device(struct udev_device *device)
 
 		if (driver == NULL)
 			return;
-	}
 
-	add_device(syspath, devname, driver, vendor, model, MODEM_TYPE_USB,
+		add_device(syspath, devname, driver, vendor, model, MODEM_TYPE_USB,
 			device, kernel_driver);
+
+		return;
+	}
+
+	//Now handle device as opposed to parent device
+	driver = udev_device_get_property_value(device, "ID_USB_DRIVER");
+	DBG("driver: %s", driver);
+
+	for (unsigned int i = 0; vendor_list[i].driver; i++) {
+		if (g_str_equal(vendor_list[i].drv, driver) == FALSE)
+			continue;
+
+		if (vendor_list[i].vid) {
+			if (!g_str_equal(vendor_list[i].vid, vendor))
+				continue;
+		}
+
+		if (vendor_list[i].pid) {
+			if (!g_str_equal(vendor_list[i].pid, model))
+				continue;
+		}
+
+		driver = vendor_list[i].driver;
+	}
+
+	if (driver == NULL)
+		return;
+
+	add_device(devsyspath, devname, driver, vendor, model, MODEM_TYPE_USB,
+		device, kernel_driver);
 }
 
 static const struct {
